### ISSUE TITLE
WordPress.Security.ValidatedSanitizedInput.InputNotValidated does not recognize `filter_var()`

### ISSUE THREAD

**Issue by Rarst** [WordPress-Coding-Standards/WordPress-Coding-Standards#1541]
## Bug Description

`$_SERVER` filtered through `filter_var()` as float is mistakenly flagged as unvalidated.

## Minimal Code Snippet
```php
$request_time = filter_var( wp_unslash( $_SERVER['REQUEST_TIME_FLOAT'] ), FILTER_VALIDATE_FLOAT );
```
## Error Code
```
 28 | ERROR | [ ] Detected usage of a non-validated input variable:
    |       |     $_SERVER
    |       |     (WordPress.Security.ValidatedSanitizedInput.InputNotValidated)
```
## Environment

| Question               | Answer
| ------------------------| -------
| PHP version             | 7.2.11 
| PHP_CodeSniffer version | 3.3.2
| WPCS version            | 1.2.0
| WPCS install type       | Composer local

## Tested Against `develop` branch?
- [ ] I have verified the issue still exists in the `develop` branch of WPCS.

**Comment by jrfnl** [comment]
Hi @Rarst I've had a look at this and this is not a false positive.

The terminology used by the sniff may be the cause of the confusion:
* "validation" in the context of this sniff means: checking if variable exists using `isset()`, `empty()` etc.
* "sanitization" in the context of this sniff means: cleaning the contents of the received data.

While `filter_var()` is a valid way to sanitize a variable, it does not _validate_ the variable and **will** throw a "Undefined index/variable" notice when passed a variable which doesn't exist.

See: https://3v4l.org/GXW62

Did you maybe intend to use `filter_input()` ? `filter_input()` will validate a variable and optionally can sanitize it too.
```php
$request_time = filter_input( INPUT_SERVER, 'REQUEST_TIME_FLOAT', FILTER_VALIDATE_FLOAT );
```

I suggest we close this issue.

**Comment by Rarst** [comment]
> "validation" in the context of this sniff means: checking if variable exists

I am not sure I agree "check if variable exist" is a general definition of validation. This code validates that input is a float or returns `false`. This remains true for undefined input, input _does_ still get validated (even though it's not defined) and result is a validation failure as it should be.

The important part here, in my opinion, is that there is no possibility of _invalid data getting through_. Which is my idea of validation process.

I would argue that if the purpose of a sniff is to exclusively check for definition then _message should say that_. "Validated" is a more complex concept with a context that is not sufficiently covered by error message as written.

> Did you maybe intend to use filter_input() ?

Unfortunately `INPUT_SERVER` is routinely unavailable in the wild, depending on server configuration, and cannot be shipped in public code (unlike `_GET`/`_POST`).

**Comment by jrfnl** [comment]
@Rarst Oh, but I totally agree that the terms are used incorrectly in the sniff. That has been that way since the sniff was added, way back before I ever got involved. That why I said "_in the context of this sniff_" in my previous reply.

So, yes, the "validation" check in the sniff is about preventing error notices, the "sanitization" check in the sniff is about what you and I would call _input validation_.

Suggestions on how to improve the error messages (without causing additional confusion) are welcome of course ;-)

**Comment by Rarst** [comment]
How about:

> Detected usage of a possibly undefined input variable: $_SERVER['REQUEST_TIME_FLOAT']

Though if variable is set by PHP itself that is still somewhat unhelpful/annoying, but it's more clear what sniff has the problem with.

**Comment by jrfnl** [comment]
@Rarst What about this:

> Detected usage of a possibly undefined superglobal array index: $_SERVER['REQUEST_TIME_FLOAT']. Use isset() or empty() to check the variable exists before using it

Might be a bit longer, but also more helpful.

**Comment by Rarst** [comment]
Good enough. :+1:

**Comment by jrfnl** [comment]
Okidoki, will pull it as part of the current series (three more to go before this one).
